In [1]:

try:
    from jyquickhelper import add_notebook_menu 
except:
    !pip install jyquickhelper
    from jyquickhelper import add_notebook_menu
    

add_notebook_menu()
#Table du contenu pour  naviger facilement le notebook. 

## Notebook presentation

This notebook is focused on presenting the benchmark results for the Rips complex, Rips edge collapse complex, and alpha complex. The evaluation is performed on the ORBIT15K dataset. The notebook includes two tests, each utilizing distinct machine learning pipelines, which will be described in subsequent steps.

## Imports 

In [2]:
import pandas as pd
import ast
import numpy as np
from memory_profiler import profile
from sklearn.preprocessing   import MinMaxScaler
from sklearn.pipeline        import Pipeline
from sklearn.svm             import SVC
from sklearn.ensemble        import RandomForestClassifier
from sklearn.neighbors       import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
from gudhi.representations import PersistenceImage
from gudhi.representations import BottleneckDistance
from gudhi.representations import Landscape
import gudhi as gd

In [3]:
pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

##  Second Test : Selection of the representation module

### Pipeline 

In [4]:
pipe = Pipeline([("Separator", gd.representations.DiagramSelector(limit=np.inf, point_type="finite")),
                    ("Scaler",    gd.representations.DiagramScaler(scalers=[([0,1], MinMaxScaler())])),
                    ("TDA",       gd.representations.PersistenceImage()),
                    ("Estimator", RandomForestClassifier())])

param =    [
                                
                {"Scaler__use":         [True],
                "TDA":                 [gd.representations.Landscape()], 
                "TDA__resolution":     [100],
                "Estimator":           [RandomForestClassifier()]},
            
                {"Scaler__use":         [True],
                "TDA":                 [gd.representations.PersistenceImage()], 
                "TDA__resolution":     [ [5,5], [6,6] ],
                "TDA__bandwidth":      [0.01, 0.1, 1.0, 10.0],
                "Estimator":           [RandomForestClassifier()]},
                {"Scaler__use":         [False],
                "TDA":                 [gd.representations.BottleneckDistance()], 
                "TDA__epsilon":        [0.1], 
                "Estimator":           [RandomForestClassifier()]}
            ]

model = GridSearchCV(pipe, param, cv=4, error_score='raise')   

### Alpha Complex :

In [5]:
df_alpha_test_1=pd.read_csv("results_alpha.csv")
df_alpha_test_2=pd.read_csv("resultsalpha_6245.csv")
df_alpha_test_3=pd.read_csv("resultsalpha_625.csv")
df_alpha_test_pipeline1=pd.concat([df_alpha_test_1,df_alpha_test_2,df_alpha_test_3])
df_alpha_test_pipeline1=df_alpha_test_pipeline1.sort_values(by="test_accuracy",ascending=False)
df_alpha_test_pipeline1 = df_alpha_test_pipeline1.reset_index(drop=True)
                
df_complex = pd.json_normalize(df_alpha_test_pipeline1["complex_parameters"].apply(ast.literal_eval))
df_complex=df_complex.reset_index(drop=True)

df_alpha_test_pipeline1 = pd.concat([df_complex, df_alpha_test_pipeline1], axis=1)
df_alpha_test_pipeline1['Estimator']=" RandomForestClassifier"
df_alpha_test_pipeline1

complex_type precision                               complex_parameters   
0        alpha     exact  {'complex_type': 'alpha', 'precision': 'exact'}  \
1        alpha     exact  {'complex_type': 'alpha', 'precision': 'exact'}   
2        alpha      fast   {'complex_type': 'alpha', 'precision': 'fast'}   
3        alpha      safe   {'complex_type': 'alpha', 'precision': 'safe'}   
4        alpha     exact  {'complex_type': 'alpha', 'precision': 'exact'}   
5        alpha      fast   {'complex_type': 'alpha', 'precision': 'fast'}   
6        alpha      safe   {'complex_type': 'alpha', 'precision': 'safe'}   
7        alpha      safe   {'complex_type': 'alpha', 'precision': 'safe'}   
8        alpha      fast   {'complex_type': 'alpha', 'precision': 'fast'}   

   elapsed_time  time for diagram computation  time for pipeline creation   
0        71.804                         9.205                         0.0  \
1        29.435                         9.576                         0.0   
2        51.927                         2.583                         0.0   
3        53.611                         2.338                         0.0   
4        39.283                        11.054                         0.0   
5        24.089                         3.178                         0.0   
6        30.043                         2.895                         0.0   
7        25.323                         2.504                         0.0   
8        22.292                         2.753                         0.0   

   time for pipeline fitting and testing  elapsed_memory   
0                                     63           1.043  \
1                                     20           0.023   
2                                     49           6.922   
3                                     51           0.176   
4                                     28           0.766   
5                                     21           5.309   
6                                     27           9.238   
7                                     23           1.141   
8                                     20          10.285   

   memory for diagram computation  memory for pipeline creation   
0                           0.000                           0.0  \
1                           0.000                           0.0   
2                           0.285                           0.0   
3                           0.000                           0.0   
4                           0.008                           0.0   
5                           1.547                           0.0   
6                           7.402                           0.0   
7                           0.000                           0.0   
8                           1.508                           0.0   

   memory for pipeline fitting and testing  test_accuracy   
0                                        1       0.619048  \
1                                        0       0.619048   
2                                        7       0.571429   
3                                        0       0.523810   
4                                        1       0.476190   
5                                        4       0.380952   
6                                        2       0.380952   
7                                        1       0.333333   
8                                        9       0.238095   

                     best parameters for the complex                Estimator  
0  {'Estimator': RandomForestClassifier(), 'Scale...   RandomForestClassifier  
1  {'Estimator': RandomForestClassifier(), 'Scale...   RandomForestClassifier  
2  {'Estimator': RandomForestClassifier(), 'Scale...   RandomForestClassifier  
3  {'Estimator': RandomForestClassifier(), 'Scale...   RandomForestClassifier  
4  {'Estimator': RandomForestClassifier(), 'Scale...   RandomForestClassifier  
5  {'Estimator': RandomForestClassifier(), 'Scale...   RandomForestClassifier  
6  {'Estimator': 

In [6]:
columns_to_keep = ['complex_type', 'precision', 'elapsed_time', 'elapsed_memory', 'test_accuracy']

# Keep only the specified columns
df_alpha_test_pipeline1_filtered = df_alpha_test_pipeline1[columns_to_keep]
df_alpha_test_pipeline1_filtered

complex_type precision  elapsed_time  elapsed_memory  test_accuracy
0        alpha     exact        71.804           1.043       0.619048
1        alpha     exact        29.435           0.023       0.619048
2        alpha      fast        51.927           6.922       0.571429
3        alpha      safe        53.611           0.176       0.523810
4        alpha     exact        39.283           0.766       0.476190
5        alpha      fast        24.089           5.309       0.380952
6        alpha      safe        30.043           9.238       0.380952
7        alpha      safe        25.323           1.141       0.333333
8        alpha      fast        22.292          10.285       0.238095

In [7]:
print(" MEAN MEMORY ",np.mean(df_alpha_test_pipeline1["elapsed_memory"]))
print(" MEAN TIME ",np.mean(df_alpha_test_pipeline1["elapsed_time"]))

for e in df_alpha_test_pipeline1['best parameters for the complex']:
    print(e)

 MEAN MEMORY  3.878111111111112
 MEAN TIME  38.64522222222222
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[6, 6]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [6, 6]}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5, 5]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [5, 5]}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[6, 6]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [6, 6]}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5

### results for alpha: 

* Best persistence representations: Persistence Image
* Best precision: exact
* Mean of elapsed_time: 38.645222
* Mean of elapsed_memory: 3.878 MB
* Best test accuracy: 0.619048
* Best elapsed time for best acc: 29.435 seconds
* Best elapsed memory for best acc: 0.023 MB


### Edge Collapse Rips Complex :

In [8]:
df_edge_test_1=pd.read_csv("results_edge.csv")
df_edge_test_2=pd.read_csv("resultsedge_6245.csv")
df_edge_test_3=pd.read_csv("resultsedge_625.csv")
df_edge_test_4=pd.read_csv("resultsedge_6255558.csv")
df_edge_test_pipeline1=pd.concat([df_edge_test_1,df_edge_test_2,df_edge_test_3,df_edge_test_4])
df_edge_test_pipeline1=df_edge_test_pipeline1.sort_values(by="test_accuracy",ascending=False)
df_edge_test_pipeline1 = df_edge_test_pipeline1.reset_index(drop=True)
                
df_complex = pd.json_normalize(df_edge_test_pipeline1["complex_parameters"].apply(ast.literal_eval))
df_complex=df_complex.reset_index(drop=True)



df_edge_test_pipeline1 = pd.concat([df_complex, df_edge_test_pipeline1], axis=1)
df_edge_test_pipeline1['Estimator']=" RandomForestClassifier"
df_edge_test_pipeline1

complex_type  max_dimension sparse  max_edge_length  nb_iterations   
0          edge              2   None         0.516584              2  \
1          edge              2   None         0.575211              3   
2          edge              2   None         0.649735              2   
3          edge              2   None         0.425860              3   
4          edge              2   None         0.439468              3   
5          edge              2   None         0.746055              3   
6          edge              2   None         0.590269              2   
7          edge              2   None         0.358031              2   
8          edge              2   None         0.284186              3   
9          edge              2   None         0.512279              2   
10         edge              2   None         0.190465              2   
11         edge              2   None         0.443120              3   
12         edge              2   None         0.516584              3   
13         edge              2   None         0.585445              3   
14         edge              2   None         0.857695              2   
15         edge              2   None         0.585445              2   
16         edge              2   None         0.851250              3   
17         edge              2   None         0.278639              2   
18         edge              2   None         0.353158              3   
19         edge              2   None         0.836042              3   
20         edge              2   None         0.726875              2   
21         edge              2   None         0.661717              3   
22         edge              2   None         0.746055              2   
23         edge              2   None         0.503620              3   
24         edge              2   None         0.661717              2   
25         edge              2   None         0.353158              2   
26         edge              2   None         0.191713              2   
27         edge              2   None         0.666974              3   
28         edge              2   None         0.431968              2   
29         edge              2   None         0.590269              3   
30         edge              2   None         0.666974              2   
31         edge              2   None         0.751931              3   
32         edge              2   None         0.364614              3   
33         edge              2   None         0.642946              3   
34         edge              2   None         0.642946              2   
35         edge              2   None         0.367327              2   
36         edge              2   None         0.275171              2   
37         edge              2   None         0.851250              2   
38         edge              2   None         0.751931              2   
39         edge              2   None         0.857695              3   
40         edge              2   None         0.439468              2   
41         edge              2   None         0.503620              2   
42         edge              2   None         0.364614              2   
43         edge              2   None         0.184333              3   
44         edge              2   None         0.286114              3   
45         edge              2   None         0.512279              3   
46         edge              2   None         0.278639              3   
47         edge              2   None         0.284186              2   
48         edge              2   None         0.425860              2   
49         edge              2   None         0.191713              3   
50         edge              2   None         0.358031              3   
51         edge              2   None         0.367327              3   
52         edge              2   None         0.732426              3   
53         edge              2   None         0.732426       

In [9]:
print(" MEAN MEMORY: ",np.mean(df_edge_test_pipeline1["elapsed_memory"]))
print(" MEAN TIME : ",np.mean(df_edge_test_pipeline1["elapsed_time"]))

for e in df_edge_test_pipeline1['best parameters for the complex']:
    print(e)

 MEAN MEMORY:  75094.70731944445
 MEAN TIME :  308.4976666666667
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[6, 6]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [6, 6]}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[6, 6]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [6, 6]}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': La

In [10]:
mask = df_edge_test_pipeline1['test_accuracy'] > 0.6
columns_to_keep = ['complex_type', 'max_edge_length','max_dimension','nb_iterations', 'elapsed_time', 'elapsed_memory', 'test_accuracy']

df_edge_test_pipeline1_filtered = df_edge_test_pipeline1[columns_to_keep]
filtered_df = df_edge_test_pipeline1_filtered[mask]
filtered_df  

complex_type  max_edge_length  max_dimension  nb_iterations  elapsed_time   
0          edge         0.516584              2              2       258.302  \
1          edge         0.575211              2              3       338.189   
2          edge         0.649735              2              2       437.601   
3          edge         0.425860              2              3       246.145   
4          edge         0.439468              2              3       151.566   
5          edge         0.746055              2              3       520.796   
6          edge         0.590269              2              2       356.636   
7          edge         0.358031              2              2       118.301   
8          edge         0.284186              2              3        63.548   
9          edge         0.512279              2              2       223.390   
10         edge         0.190465              2              2        38.442   
11         edge         0.443120              2              3       176.507   

    elapsed_memory  test_accuracy  
0        55270.480       0.714286  
1        80732.895       0.666667  
2       110552.590       0.666667  
3        43914.680       0.666667  
4        35265.723       0.666667  
5       151870.590       0.666667  
6        81101.656       0.619048  
7        24753.375       0.619048  
8        10466.676       0.619048  
9        55033.195       0.619048  
10        6416.324       0.619048  
11       34378.750       0.619048

In [11]:
 
filtered_df['max_edge_length'].value_counts()


max_edge_length
0.516584    1
0.575211    1
0.649735    1
0.425860    1
0.439468    1
0.746055    1
0.590269    1
0.358031    1
0.284186    1
0.512279    1
0.190465    1
0.443120    1
Name: count, dtype: int64

In [12]:
print(" MEAN MEMORY: ",np.mean(filtered_df["elapsed_memory"]))
print(" MEAN TIME : ",np.mean(filtered_df["elapsed_time"]))

for e in df_edge_test_pipeline1['best parameters for the complex']:
    print(e)

 MEAN MEMORY:  57479.7445
 MEAN TIME :  244.1185833333333
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[6, 6]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [6, 6]}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[6, 6]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [6, 6]}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape

### results for Edge Collapse Rips Complex : 
* Best accuracy: 0.714286
* elapsed time for best acc: 258.302	
* Best persistence representations: Landscape
* Best threshold: between 5th and 6th quantile
* Mean of elapsed_time: 1173.497 S
* Mean of elapsed_memory: 75028.74 MB
* Best number of collapse: 2 and 3 



###  Rips Complex :

In [13]:
df_rips_test_1=pd.read_csv("results_rips.csv")
df_rips_test_2=pd.read_csv("resultsrips_6245.csv")
df_rips_test_3=pd.read_csv("resultsrips_625.csv")
df_rips_test_4=pd.read_csv("resultsrips_62588.csv")
df_rips_test_pipeline1=pd.concat([df_rips_test_1,df_rips_test_2,df_rips_test_3,df_rips_test_4])
df_rips_test_pipeline1=df_rips_test_pipeline1.sort_values(by="test_accuracy",ascending=False)
df_rips_test_pipeline1 = df_rips_test_pipeline1.reset_index(drop=True)
                
df_complex = pd.json_normalize(df_rips_test_pipeline1["complex_parameters"].apply(ast.literal_eval))
df_complex=df_complex.reset_index(drop=True)

df_rips_test_pipeline1 = pd.concat([df_complex, df_rips_test_pipeline1], axis=1)
df_rips_test_pipeline1['Estimator']=" RandomForestClassifier"

df_rips_test_pipeline1

complex_type  max_dimension sparse  max_edge_length   
0          rips              2   None         0.836201  \
1          rips              2   None         0.732426   
2          rips              2   None         0.501485   
3          rips              2   None         0.277081   
4          rips              2   None         0.439468   
5          rips              2   None         0.358031   
6          rips              2   None         0.503620   
7          rips              2   None         0.190465   
8          rips              2   None         0.364614   
9          rips              2   None         0.573518   
10         rips              2   None         0.355923   
11         rips              2   None         0.851250   
12         rips              2   None         0.512279   
13         rips              2   None         0.284186   
14         rips              2   None         0.746055   
15         rips              2   None         0.649735   
16         rips              2   None         0.648507   
17         rips              2   None         0.661717   
18         rips              2   None         0.431968   
19         rips              2   None         0.278639   
20         rips              2   None         0.731654   
21         rips              2   None         0.585445   
22         rips              2   None         0.185702   
23         rips              2   None         0.575211   
24         rips              2   None         0.186441   
25         rips              2   None         0.429650   
26         rips              2   None         0.188577   
27         rips              2   None         0.836042   

                                   complex_parameters  elapsed_time   
0   {'complex_type': 'rips', 'max_dimension': 2, '...     15967.850  \
1   {'complex_type': 'rips', 'max_dimension': 2, '...     13377.622   
2   {'complex_type': 'rips', 'max_dimension': 2, '...      5325.229   
3   {'complex_type': 'rips', 'max_dimension': 2, '...      1431.260   
4   {'complex_type': 'rips', 'max_dimension': 2, '...      2859.286   
5   {'complex_type': 'rips', 'max_dimension': 2, '...      2419.382   
6   {'complex_type': 'rips', 'max_dimension': 2, '...      5469.453   
7   {'complex_type': 'rips', 'max_dimension': 2, '...       549.468   
8   {'complex_type': 'rips', 'max_dimension': 2, '...      1638.864   
9   {'complex_type': 'rips', 'max_dimension': 2, '...      7001.134   
10  {'complex_type': 'rips', 'max_dimension': 2, '...      2382.761   
11  {'complex_type': 'rips', 'max_dimension': 2, '...     20080.152   
12  {'complex_type': 'rips', 'max_dimension': 2, '...      4639.547   
13  {'complex_type': 'rips', 'max_dimension': 2, '...       862.682   
14  {'complex_type': 'rips', 'max_dimension': 2, '...     14514.367   
15  {'complex_type': 'rips', 'max_dimension': 2, '...      9997.927   
16  {'complex_type': 'rips', 'max_dimension': 2, '...      9171.576   
17  {'complex_type': 'rips', 'max_dimension': 2, '...     10141.458   
18  {'complex_type': 'rips', 'max_dimension': 2, '...      3685.818   
19  {'complex_type': 'rips', 'max_dimension': 2, '...      1542.938   
20  {'complex_type': 'rips', 'max_dimension': 2, '...     12664.431   
21  {'complex_type': 'rips', 'max_dimension': 2, '...      6536.873   
22  {'complex_type': 'rips', 'max_dimension': 2, '...       889.486   
23  {'complex_type': 'rips', 'max_dimension': 2, '...      7116.080   
24  {'complex_type': 'rips', 'max_dimension': 2, '...      1005.171   
25  {'complex_type': 'rips', 'max_dimension': 2, '...      3646.665   
26  {'complex_type': 'rips', 'max_dimension': 2, '...         8.538   
27  {'complex_type': 'rips', 'max_dimension': 2, '...     16962.920   

    time for diagram computation  time for pipeline creation   
0                      15956.725                       0.001  \
1                      13365.973                       0.000   
2                       5314.198                       0.000   
3   

In [14]:
mask = df_rips_test_pipeline1['test_accuracy'] > 0.6

# Apply the mask to the DataFrame
filtered_df = df_rips_test_pipeline1[mask]
print(np.mean(filtered_df["elapsed_memory"]))
print(np.mean(filtered_df["elapsed_time"]))
print (filtered_df['max_edge_length'].value_counts())

for e in filtered_df['best parameters for the complex']:
    print(e)


inf
5924.94375
max_edge_length
0.836201    1
0.732426    1
0.501485    1
0.277081    1
0.439468    1
0.358031    1
0.503620    1
0.190465    1
Name: count, dtype: int64
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5, 5]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [5, 5]}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5, 5]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [5, 5]}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5, 5]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [5, 5]}
{'Estimator': RandomForestClassifier(), 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[6, 6]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [6, 6]}
{'Estimator': RandomForestClassifier(), 'Scale

In [15]:

mask = df_rips_test_pipeline1['test_accuracy'] > 0.6
columns_to_keep = ['complex_type', 'max_edge_length', 'elapsed_time', 'elapsed_memory', 'test_accuracy']
df_edge_test_pipeline1_filtered = df_rips_test_pipeline1[columns_to_keep]

# Apply the mask to the DataFrame
filtered_df = df_edge_test_pipeline1_filtered[mask]
filtered_df  



complex_type  max_edge_length  elapsed_time  elapsed_memory  test_accuracy
0         rips         0.836201     15967.850             inf       0.761905
1         rips         0.732426     13377.622             inf       0.714286
2         rips         0.501485      5325.229             inf       0.666667
3         rips         0.277081      1431.260             inf       0.666667
4         rips         0.439468      2859.286             inf       0.666667
5         rips         0.358031      2419.382             inf       0.619048
6         rips         0.503620      5469.453             inf       0.619048
7         rips         0.190465       549.468             inf       0.619048

###  results for  Rips Complex :
* Best accuracy: 0.761905
* elapsed time for best acc: 15967.850	
* Best persistence representations: Persistence Image
* Best threshold: NA
* Mean of elapsed_time: 5924.94375 S
* Mean of elapsed_memory: inf MB


## Third Test : RandomForestClassifier Hyperparameters Tuning

### Pipeline: 

In [16]:
pipe = Pipeline([("Separator", gd.representations.DiagramSelector(limit=np.inf, point_type="finite")),
                    ("Scaler",    gd.representations.DiagramScaler(scalers=[([0,1], MinMaxScaler())])),
                    ("TDA",       gd.representations.PersistenceImage()),
                    ("Estimator", RandomForestClassifier())])

param =    [
                                
                {"Scaler__use":         [True],
                "TDA":                 [gd.representations.Landscape()], 
                "TDA__resolution":     [100],
                "Estimator":           [RandomForestClassifier()],
                "Estimator": [RandomForestClassifier()],
               "Estimator__n_estimators": [50, 100, 200],
                "Estimator__max_depth": [None, 5, 10],
                "Estimator__min_samples_split": [2, 5, 10],
                "Estimator__min_samples_leaf": [1, 2, 4]},
            
                {"Scaler__use":         [True],
                "TDA":                 [gd.representations.PersistenceImage()], 
                "TDA__resolution":     [ [5,5], [6,6] ],
                "TDA__bandwidth":      [0.01, 0.1, 1.0, 10.0],
                "Estimator":           [RandomForestClassifier()],
                "Estimator__n_estimators": [50, 100, 200],
                "Estimator__max_depth": [None, 5, 10],
                "Estimator__min_samples_split": [2, 5, 10],
                "Estimator__min_samples_leaf": [1, 2, 4]},
            ]

model = GridSearchCV(pipe, param, cv=4, error_score='raise')    

### Alpha Complex :

In [17]:
df_alpha_test_1=pd.read_csv("results2alpha_6245.csv")
df_alpha_test_2=pd.read_csv("results2alpha_625.csv")
df_alpha_test_3=pd.read_csv("results2alpha_62588.csv")
df_alpha_test_4=pd.read_csv("results2alpha_6255558.csv")
df_alpha_test_5=pd.read_csv("results2alpha_621238.csv")
df_alpha_test_6=pd.read_csv("results2alpha_99846.csv")
df_alpha_test_7=pd.read_csv("results2alpha_9986.csv")

df_alpha_test_pipeline1=pd.concat([df_alpha_test_1,df_alpha_test_2,df_alpha_test_3,df_alpha_test_4,df_alpha_test_5,df_alpha_test_6,df_alpha_test_7])
df_alpha_test_pipeline1=df_alpha_test_pipeline1.sort_values(by="test_accuracy",ascending=False)
df_alpha_test_pipeline1 = df_alpha_test_pipeline1.reset_index(drop=True)
                
df_complex = pd.json_normalize(df_alpha_test_pipeline1["complex_parameters"].apply(ast.literal_eval))
df_complex=df_complex.reset_index(drop=True)

df_alpha_test_pipeline1 = pd.concat([df_complex, df_alpha_test_pipeline1], axis=1)
df_alpha_test_pipeline1['Estimator']=" RandomForestClassifier"
df_alpha_test_pipeline1

complex_type precision                               complex_parameters   
0         alpha      safe   {'complex_type': 'alpha', 'precision': 'safe'}  \
1         alpha      safe   {'complex_type': 'alpha', 'precision': 'safe'}   
2         alpha     exact  {'complex_type': 'alpha', 'precision': 'exact'}   
3         alpha      safe   {'complex_type': 'alpha', 'precision': 'safe'}   
4         alpha      safe   {'complex_type': 'alpha', 'precision': 'safe'}   
5         alpha      fast   {'complex_type': 'alpha', 'precision': 'fast'}   
6         alpha      fast   {'complex_type': 'alpha', 'precision': 'fast'}   
7         alpha     exact  {'complex_type': 'alpha', 'precision': 'exact'}   
8         alpha      safe   {'complex_type': 'alpha', 'precision': 'safe'}   
9         alpha      fast   {'complex_type': 'alpha', 'precision': 'fast'}   
10        alpha     exact  {'complex_type': 'alpha', 'precision': 'exact'}   
11        alpha      safe   {'complex_type': 'alpha', 'precision': 'safe'}   
12        alpha     exact  {'complex_type': 'alpha', 'precision': 'exact'}   
13        alpha      fast   {'complex_type': 'alpha', 'precision': 'fast'}   
14        alpha     exact  {'complex_type': 'alpha', 'precision': 'exact'}   
15        alpha     exact  {'complex_type': 'alpha', 'precision': 'exact'}   
16        alpha      fast   {'complex_type': 'alpha', 'precision': 'fast'}   
17        alpha     exact  {'complex_type': 'alpha', 'precision': 'exact'}   
18        alpha      safe   {'complex_type': 'alpha', 'precision': 'safe'}   
19        alpha      fast   {'complex_type': 'alpha', 'precision': 'fast'}   
20        alpha      fast   {'complex_type': 'alpha', 'precision': 'fast'}   

    elapsed_time  time for diagram computation  time for pipeline creation   
0       1390.002                         2.451                         0.0  \
1       1522.215                         2.294                         0.0   
2       1537.387                         8.572                         0.0   
3       1527.480                         2.409                         0.0   
4       1478.673                         2.329                         0.0   
5       1470.563                         2.644                         0.0   
6       1530.352                         2.530                         0.0   
7       1161.026                         8.810                         0.0   
8       1986.921                         2.324                         0.0   
9       1562.842                         2.480                         0.0   
10      1523.878                         8.610                         0.0   
11      1538.945                         2.316                         0.0   
12      1504.587                         9.044                         0.0   
13      1538.717                         2.849                         0.0   
14      1473.822                         8.741                         0.0   
15      1391.146                         8.786                         0.0   
16      1834.864                         2.556                         0.0   
17      1474.027                         8.916                         0.0   
18      1525.255                         2.439                         0.0   
19      1467.000                         2.574                         0.0   
20      1131.558                         2.435                         0.0   

    time for pipeline fitting and testing  elapsed_memory   
0                                    1388          24.684  \
1                                    1520           4.359   
2                                    1529           2.895   
3                                    1525           2.039   
4                                    1476           2.527   
5                                    1468          15.801   
6                                    1528          13.410   
7                                    1152           0.453   
8                   

In [18]:
print(" Mean memory :",np.mean(df_alpha_test_pipeline1["elapsed_memory"]))
print(" Mean time :",np.mean(df_alpha_test_pipeline1["elapsed_time"]))

for e in df_alpha_test_pipeline1['best parameters for the complex']:
    print(e)
    print("---------------------------------------------------------------------------------------------------------------")
    print("---------------------------------------------------------------------------------------------------------------")


 Mean memory : 7.630904761904761
 Mean time : 1503.3933333333334
{'Estimator': RandomForestClassifier(), 'Estimator__max_depth': None, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 100, 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
---------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------
{'Estimator': RandomForestClassifier(max_depth=10, min_samples_leaf=2, n_estimators=50), 'Estimator__max_depth': 10, 'Estimator__min_samples_leaf': 2, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 50, 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5, 5]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [5, 5]}
---------------------------------------------------------------------------------------------------------------
----------------

In [19]:
mask = df_alpha_test_pipeline1['test_accuracy'] > 0.6

# Apply the mask to the DataFrame
filtered_df = df_alpha_test_pipeline1[mask]
print(np.mean(filtered_df["elapsed_memory"]))
print(np.mean(filtered_df["elapsed_time"]))

for e in filtered_df['best parameters for the complex']:
    print(e)




10.645999999999999
1483.2013333333332
{'Estimator': RandomForestClassifier(), 'Estimator__max_depth': None, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 100, 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(max_depth=10, min_samples_leaf=2, n_estimators=50), 'Estimator__max_depth': 10, 'Estimator__min_samples_leaf': 2, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 50, 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5, 5]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [5, 5]}
{'Estimator': RandomForestClassifier(n_estimators=50), 'Estimator__max_depth': None, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 50, 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5, 5]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [5, 5]}


In [20]:

mask = df_alpha_test_pipeline1['test_accuracy'] > 0.6
columns_to_keep = ['complex_type', 'precision', 'elapsed_time', 'elapsed_memory', 'test_accuracy']
df_alpha_test_pipeline1_filtered = df_alpha_test_pipeline1[columns_to_keep]

# Apply the mask to the DataFrame
filtered_df = df_alpha_test_pipeline1_filtered[mask]
filtered_df  

complex_type precision  elapsed_time  elapsed_memory  test_accuracy
0        alpha      safe      1390.002          24.684       0.619048
1        alpha      safe      1522.215           4.359       0.619048
2        alpha     exact      1537.387           2.895       0.619048

### results for Alpha Complex: 

* Best persistence representations: Persistence Image
* Best precision: exact
* Mean of elapsed_time: 38.645222
* Mean of elapsed_memory: 3.878 MB
* Best test accuracy: 0.619048
* Best elapsed time for best acc: 29.435 seconds


### Edge Collapse Rips Complex

In [21]:
df_edge_test_1=pd.read_csv("results2edge_6245.csv")
df_edge_test_2=pd.read_csv("results2edge_625.csv")
df_edge_test_3=pd.read_csv("results2edge_62588.csv")
df_edge_test_4=pd.read_csv("results2edge_6255558.csv")
df_edge_test_5=pd.read_csv("results2edge_76338.csv")
df_edge_test_6=pd.read_csv("results2edge_336238.csv")
df_edge_test_7=pd.read_csv("results2edge_621238.csv")
df_edge_test_8=pd.read_csv("results2edge_7644338.csv")
df_edge_test_9=pd.read_csv("results2edge_9986.csv")
df_edge_test_10=pd.read_csv("results2edge_89846.csv")
df_edge_test_11=pd.read_csv("results2edge_56238.csv")

df_edge_test_pipeline1=pd.concat([df_edge_test_1,df_edge_test_2,df_edge_test_3,df_edge_test_4,df_edge_test_5,df_edge_test_6,df_edge_test_7,df_edge_test_8,df_edge_test_9,df_edge_test_10,df_edge_test_11])
df_edge_test_pipeline1=df_edge_test_pipeline1.sort_values(by="test_accuracy",ascending=False)
df_edge_test_pipeline1 = df_edge_test_pipeline1.reset_index(drop=True)
                
df_complex = pd.json_normalize(df_edge_test_pipeline1["complex_parameters"].apply(ast.literal_eval))
df_complex=df_complex.reset_index(drop=True)



df_edge_test_pipeline1 = pd.concat([df_complex, df_edge_test_pipeline1], axis=1)
df_edge_test_pipeline1['Estimator']=" RandomForestClassifier"
df_edge_test_pipeline1

complex_type  max_dimension sparse  max_edge_length  nb_iterations   
0           edge              2   None         0.358031              3  \
1           edge              2   None         0.661717              2   
2           edge              2   None         0.585445              3   
3           edge              2   None         0.496333              3   
4           edge              2   None         0.668441              2   
5           edge              2   None         0.671234              3   
6           edge              2   None         0.364614              3   
7           edge              2   None         0.353158              3   
8           edge              2   None         0.726875              2   
9           edge              2   None         0.726875              3   
10          edge              2   None         0.503620              3   
11          edge              2   None         0.836201              2   
12          edge              2   None         0.671234              2   
13          edge              2   None         0.355923              3   
14          edge              2   None         0.284186              2   
15          edge              2   None         0.858712              2   
16          edge              2   None         0.512279              3   
17          edge              2   None         0.858712              3   
18          edge              2   None         0.501485              2   
19          edge              2   None         0.277081              3   
20          edge              2   None         0.517931              3   
21          edge              2   None         0.741360              3   
22          edge              2   None         0.731654              2   
23          edge              2   None         0.732426              3   
24          edge              2   None         0.575211              3   
25          edge              2   None         0.579512              3   
26          edge              2   None         0.503620              2   
27          edge              2   None         0.844886              2   
28          edge              2   None         0.370012              3   
29          edge              2   None         0.580863              2   
30          edge              2   None         0.746055              2   
31          edge              2   None         0.446094              3   
32          edge              2   None         0.436307              3   
33          edge              2   None         0.738313              3   
34          edge              2   None         0.731654              3   
35          edge              2   None         0.436307              2   
36          edge              2   None         0.361818              3   
37          edge              2   None         0.842572              2   
38          edge              2   None         0.642946              2   
39          edge              2   None         0.654961              2   
40          edge              2   None         0.863527              2   
41          edge              2   None         0.863527              3   
42          edge              2   None         0.501485              3   
43          edge              2   None         0.446094              2   
44          edge              2   None         0.593858              2   
45          edge              2   None         0.358031              2   
46          edge              2   None         0.355923              2   
47          edge              2   None         0.190465              2   
48          edge              2   None         0.732426              2   
49          edge              2   None         0.505314              3   
50          edge              2   None         0.746055              3   
51          edge              2   None         0.278639              3   
52          edge              2   None         0.841716              3   
53      

In [22]:
print(np.mean(df_edge_test_pipeline1["elapsed_memory"]))
print(np.mean(df_edge_test_pipeline1["elapsed_time"]))

for e in df_edge_test_pipeline1['best parameters for the complex']:
    print(e)

75028.74947474748
1173.6455454545455
{'Estimator': RandomForestClassifier(max_depth=10, n_estimators=200), 'Estimator__max_depth': 10, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 200, 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(max_depth=5, min_samples_split=5), 'Estimator__max_depth': 5, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 5, 'Estimator__n_estimators': 100, 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[6, 6]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [6, 6]}
{'Estimator': RandomForestClassifier(max_depth=10, min_samples_leaf=2), 'Estimator__max_depth': 10, 'Estimator__min_samples_leaf': 2, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 100, 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5, 5]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [5, 5]}
{'Estimator': RandomForestClassifier(m

In [23]:
mask = df_edge_test_pipeline1['test_accuracy'] > 0.6

# Apply the mask to the DataFrame
filtered_df = df_edge_test_pipeline1[mask]
print("Mean Memory: ",np.mean(filtered_df["elapsed_memory"]))
print("Mean Time: ", np.mean(filtered_df["elapsed_time"]))

for e in filtered_df['best parameters for the complex']:
    print(e)




Mean Memory:  95876.1323548387
Mean Time:  1262.4036774193544
{'Estimator': RandomForestClassifier(max_depth=10, n_estimators=200), 'Estimator__max_depth': 10, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 200, 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(max_depth=5, min_samples_split=5), 'Estimator__max_depth': 5, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 5, 'Estimator__n_estimators': 100, 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[6, 6]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [6, 6]}
{'Estimator': RandomForestClassifier(max_depth=10, min_samples_leaf=2), 'Estimator__max_depth': 10, 'Estimator__min_samples_leaf': 2, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 100, 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5, 5]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [5, 5]}
{'Estimator':

In [24]:
mask = df_edge_test_pipeline1['test_accuracy'] > 0.6
columns_to_keep = ['complex_type', 'max_edge_length','nb_iterations', 'elapsed_time', 'elapsed_memory', 'test_accuracy']

# Keep only the specified columns
df_edge_test_pipeline1_filtered = df_edge_test_pipeline1[columns_to_keep]
# Apply the mask to the DataFrame
filtered_df = df_edge_test_pipeline1_filtered[mask]
filtered_df

complex_type  max_edge_length  nb_iterations  elapsed_time  elapsed_memory   
0          edge         0.358031              3       924.289       24812.172  \
1          edge         0.661717              2      1236.627      112081.469   
2          edge         0.585445              3      1172.248       80295.172   
3          edge         0.496333              3      1293.219       62795.770   
4          edge         0.668441              2      1596.168      113415.434   
5          edge         0.671234              3      1233.926      116561.141   
6          edge         0.364614              3       926.261       20598.227   
7          edge         0.353158              3      1182.370       29931.305   
8          edge         0.726875              2      1414.106      155012.613   
9          edge         0.726875              3      1555.586      154611.320   
10         edge         0.503620              3      1034.109       57096.082   
11         edge         0.836201              2      1457.967      198953.895   
12         edge         0.671234              2      1237.675      116556.375   
13         edge         0.355923              3       969.423       24104.117   
14         edge         0.284186              2       889.316       10451.992   
15         edge         0.858712              2      1885.815      205436.926   
16         edge         0.512279              3      1095.164       55036.301   
17         edge         0.858712              3      1880.667      205246.816   
18         edge         0.501485              2      1051.579       57208.250   
19         edge         0.277081              3       935.718        9664.074   
20         edge         0.517931              3      1354.568       54687.297   
21         edge         0.741360              3      1359.259      153521.730   
22         edge         0.731654              2      1341.635      149649.023   
23         edge         0.732426              3      1300.524      148420.855   
24         edge         0.575211              3      1107.605       80750.793   
25         edge         0.579512              3      1156.686       83584.953   
26         edge         0.503620              2      1035.554       57096.766   
27         edge         0.844886              2      1817.183      192692.906   
28         edge         0.370012              3       922.653       17419.137   
29         edge         0.580863              2      1425.537       72631.516   
30         edge         0.746055              2      1341.077      151835.676   

    test_accuracy  
0        0.761905  
1        0.761905  
2        0.761905  
3        0.761905  
4        0.666667  
5        0.666667  
6        0.666667  
7        0.666667  
8        0.666667  
9        0.666667  
10       0.666667  
11       0.666667  
12       0.666667  
13       0.666667  
14       0.666667  
15       0.619048  
16       0.619048  
17       0.619048  
18       0.619048  
19       0.619048  
20       0.619048  
21       0.619048  
22       0.619048  
23       0.619048  
24       0.619048  
25       0.619048  
26       0.619048  
27       0.619048  
28       0.619048  
29       0.619048  
30       0.619048

{'Estimator': RandomForestClassifier(max_depth=5), 'Estimator__max_depth': 5, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 100, 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}

In [25]:
filtered_df['max_edge_length'].value_counts(), filtered_df['nb_iterations'].value_counts()



(max_edge_length
 0.858712    2
 0.671234    2
 0.726875    2
 0.503620    2
 0.501485    1
 0.580863    1
 0.370012    1
 0.844886    1
 0.579512    1
 0.575211    1
 0.732426    1
 0.731654    1
 0.741360    1
 0.517931    1
 0.277081    1
 0.358031    1
 0.512279    1
 0.661717    1
 0.284186    1
 0.355923    1
 0.836201    1
 0.353158    1
 0.364614    1
 0.668441    1
 0.496333    1
 0.585445    1
 0.746055    1
 Name: count, dtype: int64,
 nb_iterations
 3    18
 2    13
 Name: count, dtype: int64)

### results for Edge Collapse Rips Complex: 
* Best persistence representations: Persistence Image // Landscape
* Best precision: exact
* Mean of elapsed_time: 1156.59575
* Mean of elapsed_memory: 69996.14 MB
* Best test accuracy: 0.762
* Best elapsed time for best acc:924.289 seconds
* Best number of iterations: 3


### Rips Complex :

In [26]:
df_rips_test_1=pd.read_csv("results2rips_6245.csv")
df_rips_test_2=pd.read_csv("results2rips_625.csv")
df_rips_test_3=pd.read_csv("results2rips_62588.csv")
df_rips_test_4=pd.read_csv("results2rips_6255558.csv")
df_rips_test_5=pd.read_csv("results2rips_89846.csv")
df_rips_test_6=pd.read_csv("results2rips_336238.csv")
df_rips_test_7=pd.read_csv("results2rips_621238.csv")
df_rips_test_8=pd.read_csv("results2rips_7644338.csv")
df_rips_test_9=pd.read_csv("results2rips_99846.csv")
df_rips_test_10=pd.read_csv("results2rips_89846.csv")
df_rips_test_10=pd.read_csv("results2rips_336238.csv")


df_rips_test_pipeline1=pd.concat([df_rips_test_1,df_rips_test_2,df_rips_test_3,df_rips_test_4,df_rips_test_5,df_rips_test_6,df_rips_test_7,df_rips_test_8,df_rips_test_9,df_rips_test_10])
df_rips_test_pipeline1=df_rips_test_pipeline1.sort_values(by="test_accuracy",ascending=False)
df_rips_test_pipeline1 = df_rips_test_pipeline1.reset_index(drop=True)
                
df_complex = pd.json_normalize(df_rips_test_pipeline1["complex_parameters"].apply(ast.literal_eval))
df_complex=df_complex.reset_index(drop=True)



df_rips_test_pipeline1 = pd.concat([df_complex, df_rips_test_pipeline1], axis=1)
df_rips_test_pipeline1['Estimator']=" RandomForestClassifier"
df_rips_test_pipeline1

complex_type  max_dimension sparse  max_edge_length   
0          rips              2   None         0.431968  \
1          rips              2   None         0.282771   
2          rips              2   None         0.591743   
3          rips              2   None         0.364614   
4          rips              2   None         0.661717   
5          rips              2   None         0.836042   
6          rips              2   None         0.275171   
7          rips              2   None         0.362385   
8          rips              2   None         0.756757   
9          rips              2   None         0.288394   
10         rips              2   None         0.288394   
11         rips              2   None         0.429650   
12         rips              2   None         0.503620   
13         rips              2   None         0.756757   
14         rips              2   None         0.746055   
15         rips              2   None         0.567462   
16         rips              2   None         0.863527   
17         rips              2   None         0.863527   
18         rips              2   None         0.585445   
19         rips              2   None         0.512279   
20         rips              2   None         0.439468   
21         rips              2   None         0.508263   
22         rips              2   None         0.190465   
23         rips              2   None         0.446094   
24         rips              2   None         0.446094   
25         rips              2   None         0.753060   
26         rips              2   None         0.278639   
27         rips              2   None         0.439020   
28         rips              2   None         0.847134   
29         rips              2   None         0.370012   
30         rips              2   None         0.741360   
31         rips              2   None         0.519768   
32         rips              2   None         0.519768   
33         rips              2   None         0.671234   
34         rips              2   None         0.370012   
35         rips              2   None         0.435037   
36         rips              2   None         0.738313   
37         rips              2   None         0.852741   
38         rips              2   None         0.284282   
39         rips              2   None         0.284186   
40         rips              2   None         0.355923   
41         rips              2   None         0.832011   
42         rips              2   None         0.726875   
43         rips              2   None         0.648507   
44         rips              2   None         0.671234   
45         rips              2   None         0.360889   
46         rips              2   None         0.654961   
47         rips              2   None         0.851250   
48         rips              2   None         0.501485   
49         rips              2   None         0.573518   
50         rips              2   None         0.188862   
51         rips              2   None         0.731654   
52         rips              2   None         0.277081   
53         rips              2   None         0.353158   
54         rips              2   None         0.425860   
55         rips              2   None         0.649735   
56         rips              2   None         0.642946   
57         rips              2   None         0.858712   
58         rips              2   None         0.191933   
59         rips              2   None         0.575211   
60         rips              2   None         0.185702   
61         rips              2   None         0.661309   
62         rips              2   None         0.584920   
63         rips              2   None         0.444201   
64         rips              2   None         0.746140   
65         rips              2   None         0.368151   
66         rips              2   None         0.836201   
67         rips              2   None         0.593858   
6

In [27]:
print("Mean Memory:",np.mean(df_rips_test_pipeline1["elapsed_memory"]))
print("Mean Time :",np.mean(df_rips_test_pipeline1["elapsed_time"]))

for e in df_rips_test_pipeline1['best parameters for the complex']:
    print(e)


Mean Memory: inf
Mean Time : 7117.8712444444445
{'Estimator': RandomForestClassifier(), 'Estimator__max_depth': None, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 100, 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Estimator__max_depth': None, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 100, 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5, 5]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [5, 5]}
{'Estimator': RandomForestClassifier(max_depth=10, n_estimators=50), 'Estimator__max_depth': 10, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 50, 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[6, 6]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [6, 6]}
{'Estimator': RandomForestClassifier(min_samples_leaf=2), 'Estimator__max_depth': None, 

In [28]:
mask = df_rips_test_pipeline1['test_accuracy'] > 0.6

# Apply the mask to the DataFrame
filtered_df = df_rips_test_pipeline1[mask]
print("Mean Memory: ",np.mean(filtered_df["elapsed_memory"]))
print("Mean Time: ", np.mean(filtered_df["elapsed_time"]))
print(filtered_df["max_edge_length"].value_counts())
for e in filtered_df['best parameters for the complex']:
    print(e)




Mean Memory:  inf
Mean Time:  6916.835533333334
max_edge_length
0.756757    2
0.288394    2
0.431968    1
0.282771    1
0.591743    1
0.364614    1
0.661717    1
0.836042    1
0.275171    1
0.362385    1
0.429650    1
0.503620    1
0.746055    1
Name: count, dtype: int64
{'Estimator': RandomForestClassifier(), 'Estimator__max_depth': None, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 100, 'Scaler__use': True, 'TDA': Landscape(), 'TDA__resolution': 100}
{'Estimator': RandomForestClassifier(), 'Estimator__max_depth': None, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators': 100, 'Scaler__use': True, 'TDA': PersistenceImage(bandwidth=0.1, resolution=[5, 5]), 'TDA__bandwidth': 0.1, 'TDA__resolution': [5, 5]}
{'Estimator': RandomForestClassifier(max_depth=10, n_estimators=50), 'Estimator__max_depth': 10, 'Estimator__min_samples_leaf': 1, 'Estimator__min_samples_split': 2, 'Estimator__n_estimators'

In [29]:
mask = df_rips_test_pipeline1['test_accuracy'] > 0.6
columns_to_keep = ['complex_type', 'max_edge_length', 'elapsed_time', 'elapsed_memory', 'test_accuracy']
df_rips_test_pipeline1_filtered = df_rips_test_pipeline1[columns_to_keep]

# Apply the mask to the DataFrame
filtered_df = df_rips_test_pipeline1_filtered[mask]
filtered_df  


complex_type  max_edge_length  elapsed_time  elapsed_memory  test_accuracy
0          rips         0.431968      4051.952             inf       0.714286
1          rips         0.282771      1793.627             inf       0.714286
2          rips         0.591743      8796.108             inf       0.666667
3          rips         0.364614      2523.064             inf       0.666667
4          rips         0.661717     11925.929             inf       0.666667
5          rips         0.836042     16397.743             inf       0.666667
6          rips         0.275171      2089.808             inf       0.619048
7          rips         0.362385      2592.378             inf       0.619048
8          rips         0.756757     12063.922             inf       0.619048
9          rips         0.288394      1321.127             inf       0.619048
10         rips         0.288394      1321.127             inf       0.619048
11         rips         0.429650      4477.655             inf       0.619048
12         rips         0.503620      5673.698             inf       0.619048
13         rips         0.756757     12063.922             inf       0.619048
14         rips         0.746055     16660.473             inf       0.619048

### results for Edge Collapse Rips Complex: 
* Best persistence representations: Persistence Image // Landscape
* Best precision: exact
* Mean of elapsed_time: 1156.59575
* Mean of elapsed_memory: 69996.14 MB
* Best test accuracy: 0.762
* Best elapsed time for best acc:924.289 seconds
* Best number of iterations: 3
